apps-rendering/src/client/callouts.ts

```diff
@@ -0,0 +1,353 @@
+import {
+	andThen,
+	arrayParser,
+	booleanParser,
+	documentFragmentParser,
+	fieldParser,
+	map,
+	map2,
+	map7,
+	maybe,
+	numberParser,
+	parse,
+	stringParser,
+	succeed,
+} from 'parser';
+import type { Parser } from 'parser';
+import { createElement as h } from 'react';
+import ReactDOM from 'react-dom';
+import { Result } from 'result';
+import CalloutComponent from '../components/Callout';
+
+type FormData = Record<string, string | string[]>;
+
+type FormElementType =
+	| HTMLInputElement
+	| HTMLTextAreaElement
+	| HTMLSelectElement;
+
+function readFile(file: Blob): Promise<string> {
+	return new Promise((resolve, reject) => {
+		const reader = new FileReader();
+		setTimeout(reject, 30000);
+
+		reader.addEventListener('load', () => {
+			if (reader.result) {
+				const fileAsBase64 = reader.result
+					.toString()
+					.split(';base64,')[1];
+				resolve(fileAsBase64);
+			}
+		});
+
+		reader.addEventListener('error', () => {
+			reject();
+		});
+
+		reader.readAsDataURL(file);
+	});
+}
+function submitCallout(body: FormData, form: Element): void {
+	fetch(
+		'https://callouts.code.dev-guardianapis.com/formstack-campaign/submit',
+		{
+			method: 'POST',
+			body: JSON.stringify(body),
+		},
+	)
+		.then((res) => {
+			if (res.ok) {
+				form.classList.add('js-callout--success');
+			} else {
+				form.classList.add('js-callout--failure');
+			}
+		})
+		.catch(() => {
+			form.classList.add('js-callout--failure');
+		});
+}
+
+const getValueFromElement = async (
+	element: FormElementType,
+	acc: FormData,
+): Promise<FormData> => {
+	const { type, name, value } = element;
+	if (type === 'radio') {
+		const radioElement = element as HTMLInputElement;
+		if (radioElement.checked) {
+			acc[name] = value;
+		}
+	} else if (type === 'checkbox') {
+		const checkboxElement = element as HTMLInputElement;
+		if (checkboxElement.checked) {
+			acc[name]
+				? (acc[name] = [...acc[name], value])
+				: (acc[name] = [value]);
+		}
+	} else if (type === 'file') {
+		const fileElement = element as HTMLInputElement;
+		if (fileElement.files?.length) {
+			acc[name] = await readFile(fileElement.files[0]);
+		}
+	} else if (value) {
+		acc[name] = value;
+	}
+	return Promise.resolve(acc);
+};
+
+const validate = (elements: FormElementType[]): boolean => {
+	let hasError = false;
+	elements.forEach((element) => {
+		if (element.type === 'checkbox') {
+			// this shouldn't be on document??? Should be on the form
+			const mandatoryCheckboxGroup = document.getElementById(
+				`checkbox-group-${element.name}--mandatory`,
+			);
+			const checkedCheckboxes =
+				mandatoryCheckboxGroup?.querySelectorAll(':checked');
+			if (mandatoryCheckboxGroup && !checkedCheckboxes?.length) {
+				document
+					.getElementById(element.name)
+					?.classList.add('callout-field--failure');
+				hasError = true;
+			} else {
+				document
+					.getElementById(element.name)
+					?.classList.remove('callout-field--failure');
```

Comment:
Yeah good spot - I probably won't do so now though because i'm working on a PR to replace this vanillaJS with react since we hydrate the callout anyway!